### ISSUE TITLE
Custom configuration "Add request-level properties for all telemetry items" does not work?

### ISSUE THREAD

**Issue by lexdenotter** [Microsoft/ApplicationInsights-aspnetcore#873]
The example in the wiki does not work, 
`Microsoft.AspNetCore.Http.IHttpContextAccessor.HttpContext.get returned null.` 
is raised on startup. 

this is the link to the wiki:
[https://github.com/Microsoft/ApplicationInsights-aspnetcore/wiki/Custom-Configuration#add-request-level-properties-for-all-telemetry-items](wiki) 
this is my code: 

` public class PropertyTelemetryInitializer : ITelemetryInitializer
    {
        IHttpContextAccessor httpContextAccessor;

        public PropertyTelemetryInitializer(IHttpContextAccessor httpContextAccessor)
        {
            this.httpContextAccessor = httpContextAccessor;
        }

        public void Initialize(ITelemetry telemetry)
        {
            telemetry.Context.GlobalProperties["statusCode"] = httpContextAccessor.HttpContext.Response.StatusCode.ToString();
        }
    }
`
on startup in `ConfigureServices`:
`
 services.AddSingleton<ITelemetryInitializer, PropertyTelemetryInitializer>();
 services.AddApplicationInsightsTelemetry();
`

**Comment by cijothomas** [comment]
@lexdenotter Can you describe what do you mean by not working ? is your telemetry initializer not called?

**Comment by pharring** [comment]
You probably need to add a null check for `httpContextAccessor.HttpContext` because the initializer can be called outside the context of a request (e.g. during startup)

**Comment by cberthold** [comment]
I'm seeing the same thing and we are trying to solve a similar problem this very minute.  For us, it seems it is an order of operations problem.  My understanding is that the Context for each telemetry item is a deep clone off of the "Initialized" value.  We are wanting to add some correlated data to the operation telemetry that occurs after the DiagnosticListener has already occurred, maybe in a middleware downstream, after the initializer has been called and created the context.  If you change the TelemetryClient context then it will not show up until something else gets started.  This will not happen if there are no dependent calls.

To try and get around this we tried making our own ContextProvider that we store as a raw object on the OperationContext and is based on AsyncLocal similar to HttpContextAccessor.  We then call that object in the processor to get the data off as a dictionary.  The issue for us is, not everything is async/await compatible and async is not ThreadLocal compatible.  It works in some cases but not all.

**Comment by cberthold** [comment]
Also wanted to note that I think I have a possible solution I'm going to try in the morning.  I had just noticed that the RequestTelemetry is added as a feature to the request.  In theory, this should solve the immediate problem and I can now add properties to the request at any point in the pipeline.

`var telemetry = HttpContext?.Features.Get<RequestTelemetry>();
if (telemetry != null)
{
    telemetry.Context.GlobalProperties["key"] = "value";
}`

**Comment by pharring** [comment]
That's right. See, for example, what's going on in TelemetryInitializerBase.cs: https://github.com/microsoft/ApplicationInsights-aspnetcore/blob/develop/src/Microsoft.ApplicationInsights.AspNetCore/TelemetryInitializers/TelemetryInitializerBase.cs#L35

**Comment by cberthold** [comment]
Awesome.  I think that at least gives me enough of an idea how to hook it into our other pipelines.  The example, however, is incorrect or at least I'm not seeing how it would be possible for it to work.  There would never be anything in `HttpContext.Items[]`because it occurs so close to the `DefaultHttpContext` being built up (empty).  I feel like the typical scenario of what we're asking for is closer to what people would want to do with it.

**Comment by lexdenotter** [comment]
> You probably need to add a null check for `httpContextAccessor.HttpContext` because the initializer can be called outside the context of a request (e.g. during startup)

Thanks, that helped! 
Maybe it is best to update the example to an working example?
